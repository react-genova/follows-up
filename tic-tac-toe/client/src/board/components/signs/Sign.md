### With values

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
</div>
```

### Without values (hover)

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"120px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```

### A little bigger and mixed

```jsx
const { SIGN_O, SIGN_X } = require('../../modules/types/signs.constants.js');
<div style={{ display:"flex", width:"480px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <Sign hasValue type={SIGN_X} />
    <Sign hasValue type={SIGN_O} />
    <Sign type={SIGN_X} />
    <Sign type={SIGN_O} />
</div>
```