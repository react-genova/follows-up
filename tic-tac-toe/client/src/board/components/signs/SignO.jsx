import React from 'react';
import PropTypes from 'prop-types';

const SignO = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <ellipse cx="50" cy="50" rx="46" ry="46" fill="transparent" stroke={ghost ? 'lightgray' : 'yellow'} strokeWidth="8" />
    </svg>
);

SignO.propTypes = {
    ghost: PropTypes.bool,
};

SignO.defaultProps = {
    ghost: false,
};

export default SignO;
