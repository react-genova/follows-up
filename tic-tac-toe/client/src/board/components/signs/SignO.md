### On light background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px" }}>
    <SignO />
    <SignO ghost />
</div>
```

### On dark background

```jsx
<div style={{ display:"flex", width:"100px", height:"50px", padding:"2px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```

### A little bigger

```jsx
<div style={{ display:"flex", width:"200px", height:"100px", padding:"5px", backgroundColor: "rgb(50,50,50)" }}>
    <SignO />
    <SignO ghost />
</div>
```