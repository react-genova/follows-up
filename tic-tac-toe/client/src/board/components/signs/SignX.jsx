import React from 'react';
import PropTypes from 'prop-types';

const SignX = ({ ghost }) => (
    <svg width="100%" height="100%" viewBox="0 0 100 100">
        <line x1="0" y1="0" x2="100" y2="100" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
        <line x1="0" y1="100" x2="100" y2="0" stroke={ghost ? 'lightgray' : 'cyan'} strokeWidth="8" />
    </svg>
);

SignX.propTypes = {
    ghost: PropTypes.bool,
};

SignX.defaultProps = {
    ghost: false,
};


export default SignX;
