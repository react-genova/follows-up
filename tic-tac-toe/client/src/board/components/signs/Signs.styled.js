import styled from 'styled-components';

const SignContainer = styled.div`
    width: 100%;
    height: 100%;
`;

export default SignContainer;
