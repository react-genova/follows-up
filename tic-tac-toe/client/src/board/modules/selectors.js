import { createSelector } from 'reselect';

let getBoardRoot = state => state;

export const setBoardRoot = (getter) => {
    getBoardRoot = getter;
};

export const getBoard = state => getBoardRoot(state);

export const getBoardValues = createSelector([getBoard], board => board.get('values').toJS());
