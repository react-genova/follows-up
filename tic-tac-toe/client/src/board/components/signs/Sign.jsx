import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import * as SIGNS from '../../modules/types/signs.constants';
import SignContainer from './Signs.styled';
import SignO from './SignO';
import SignX from './SignX';

class Sign extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            hover: false,
        }
    }

    onMouseEnter = () => this.setState({ hover: true });

    onMouseLeave = () => this.setState({ hover: false });

    onClick = () => this.props.onClick(this.props.type);

    render() {
        const { type, hasValue } = this.props;
        const { hover } = this.state;
        return (
            <SignContainer onClick={this.onClick} onMouseEnter={this.onMouseEnter} onMouseLeave={this.onMouseLeave}>
            {
                (hasValue || hover) && {
                    [SIGNS.SIGN_O]: <SignO ghost={!hasValue} />,
                    [SIGNS.SIGN_X]: <SignX ghost={!hasValue} />,
                }[type]
            }
            </SignContainer>
        );
    }
}

Sign.propTypes = {
    hasValue: PropTypes.bool,
    onClick: PropTypes.func,
    type: PropTypes.oneOf([SIGNS.SIGN_O, SIGNS.SIGN_X]),
};

Sign.defaultProps = {
    hasValue: false,
    onClick: () => null,
    type: SIGNS.SIGN_O,
};

export default Sign;
