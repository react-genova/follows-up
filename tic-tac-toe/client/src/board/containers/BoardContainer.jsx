import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Grid from '../components/grid/Grid';
import { getPlayer1Symbol } from '../../settings/modules/selectors';
import { getBoardValues } from '../modules/selectors';
import { addBoardMove } from '../modules/action.creators';
import { valuesPropTypes, playingTypePropTypes } from '../components/grid/Grid.types';
import { isGameStarted } from '../../game/engine/selectors';

class BoardContainer extends Component {
    onCellClick = (index) => {
        const { values, playingType, gameStarted } = this.props;
        if (gameStarted && !values[index].valid) {
            const { addBoardMove: fireAddBoardMove } = this.props;
            fireAddBoardMove(index, playingType);
        }
    }

    render() {
        const { values, playingType } = this.props;
        return (<Grid values={values} playingType={playingType} onCellClick={this.onCellClick} />);
    }
}

BoardContainer.propTypes = {
    addBoardMove: PropTypes.func,
    playingType: playingTypePropTypes.isRequired,
    values: valuesPropTypes.isRequired,
    gameStarted: PropTypes.bool.isRequired,
};

BoardContainer.defaultProps = {
    addBoardMove: undefined,
};

export const mapStateToProps = state => ({
    values: getBoardValues(state),
    playingType: getPlayer1Symbol(state),
    gameStarted: isGameStarted(state),
});

export const mapDispatchToProps = {
    addBoardMove,
};

export { BoardContainer };
export default connect(mapStateToProps, mapDispatchToProps)(BoardContainer);
