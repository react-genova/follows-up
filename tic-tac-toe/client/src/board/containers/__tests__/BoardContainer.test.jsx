import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { BoardContainer, mapStateToProps, mapDispatchToProps } from '../BoardContainer';
import { getPlayer1Symbol } from '../../../settings/modules/selectors';
import { getBoardValues } from '../../modules/selectors';
import { addBoardMove } from '../../modules/action.creators';
import { SIGN_X, SIGN_O, SIGN_NONE } from '../../modules/types/signs.constants';
import { isGameStarted } from '../../../game/engine/selectors';

jest.mock('../../../settings/modules/selectors');
jest.mock('../../modules/selectors');
jest.mock('../../modules/action.creators');
jest.mock('../../../game/engine/selectors');

describe('Board conatiner', () => {
    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
        getBoardValues.mockImplementation(() => []);
    });

    it('creates valid props from state', () => {
        // setting up mocks
        const STATE = 'I am state';
        const PLAYER_1_SYMBOL = 'EX';
        const BOARD_VALUES = ['EX', 'EX', 'CIRCLE', 'EX'];
        getPlayer1Symbol.mockImplementation(() => PLAYER_1_SYMBOL);
        getBoardValues.mockImplementation(() => BOARD_VALUES);
        isGameStarted.mockImplementation(() => true);
        // invoking the method to test
        const props = mapStateToProps(STATE);
        // check result
        expect(getPlayer1Symbol).toHaveBeenCalledTimes(1);
        expect(getBoardValues).toHaveBeenCalledTimes(1);
        expect(getPlayer1Symbol).toHaveBeenCalledWith(STATE);
        expect(getBoardValues).toHaveBeenCalledWith(STATE);
        expect(props).toEqual({
            values: BOARD_VALUES,
            playingType: PLAYER_1_SYMBOL,
            gameStarted: true,
        });
    });

    it('creates valid props from actions', () => {
        expect(mapDispatchToProps).toEqual({
            addBoardMove,
        });
    });

    it('renders without crashing and fires correct moving action when clicking on an empty cell', () => {
        const PLAYER_1_SYMBOL = SIGN_X;
        const BOARD_VALUES = [
            { value: SIGN_X, valid: true },
            { value: SIGN_O, valid: true },
            { value: SIGN_X, valid: true },
            { value: SIGN_NONE, valid: false },
        ];
        const addBoardMoveMocked = jest.fn();
        const { container } = render(
            <BoardContainer gameStarted values={BOARD_VALUES} playingType={PLAYER_1_SYMBOL} addBoardMove={addBoardMoveMocked} />,
        );
        expect(container.firstChild).toMatchSnapshot();
        // click on cell with valid value => no action fired
        fireEvent.click(container.firstChild.firstChild.firstChild);
        // click on cell with not valid value => action fired
        fireEvent.click(container.firstChild.lastChild.firstChild);
        expect(addBoardMoveMocked).toHaveBeenCalledTimes(1);
        expect(addBoardMoveMocked).toHaveBeenCalledWith(3, PLAYER_1_SYMBOL);
    });

    it('down not fire any moving action if game is not started', () => {
        const PLAYER_1_SYMBOL = SIGN_X;
        const BOARD_VALUES = [
            { value: SIGN_X, valid: true },
            { value: SIGN_O, valid: true },
            { value: SIGN_X, valid: true },
            { value: SIGN_NONE, valid: false },
        ];
        const addBoardMoveMocked = jest.fn();
        isGameStarted.mockImplementation(() => false);
        const { container } = render(
            <BoardContainer
                gameStarted={false}
                values={BOARD_VALUES}
                playingType={PLAYER_1_SYMBOL}
                addBoardMove={addBoardMoveMocked}
            />,
        );
        // click on cell with not valid value => action should have been fired, when game is started
        fireEvent.click(container.firstChild.lastChild.firstChild);
        expect(addBoardMoveMocked).toHaveBeenCalledTimes(0);
    });
});
