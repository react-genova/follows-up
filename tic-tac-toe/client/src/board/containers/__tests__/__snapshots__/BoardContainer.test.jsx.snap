// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board conatiner renders without crashing and fires correct moving action when clicking on an empty cell 1`] = `
<div
  style="width: 800px; height: 800px;"
>
  <div
    class="sc-bdVaJa cpgzAv"
  >
    <div
      class="sc-bwzfXH evblgu"
      type="MOVING_SYMBOL_X"
    >
      <div
        class="sc-htpNat emEaka"
      >
        <svg
          height="100%"
          viewBox="0 0 100 100"
          width="100%"
        >
          <line
            stroke="cyan"
            stroke-width="8"
            x1="0"
            x2="100"
            y1="0"
            y2="100"
          />
          <line
            stroke="cyan"
            stroke-width="8"
            x1="0"
            x2="100"
            y1="100"
            y2="0"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-bwzfXH kbZGzz"
      type="MOVING_SYMBOL_O"
    >
      <div
        class="sc-htpNat emEaka"
      >
        <svg
          height="100%"
          viewBox="0 0 100 100"
          width="100%"
        >
          <ellipse
            cx="50"
            cy="50"
            fill="transparent"
            rx="46"
            ry="46"
            stroke="yellow"
            stroke-width="8"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-bwzfXH evblgu"
      type="MOVING_SYMBOL_X"
    >
      <div
        class="sc-htpNat emEaka"
      >
        <svg
          height="100%"
          viewBox="0 0 100 100"
          width="100%"
        >
          <line
            stroke="cyan"
            stroke-width="8"
            x1="0"
            x2="100"
            y1="0"
            y2="100"
          />
          <line
            stroke="cyan"
            stroke-width="8"
            x1="0"
            x2="100"
            y1="100"
            y2="0"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-bwzfXH evblgu"
      type="NONE"
    >
      <div
        class="sc-htpNat emEaka"
      />
    </div>
  </div>
</div>
`;
